### PR TITLE
[Datastore] Extend Azure blob to support other auth methods

### PR DESCRIPTION
**Pull request by hayesgb**
Currently, the only supported authentication methods against AzureBlobStore is AZURE_STORAGE_CONNECTION_STRING. , and possibly AZURE_STORAGE_KEY.  This enables other authentication methods including use of ServicePrincipals and SAS tokens.

### CODE REVIEW

**Comment by hayesgb** [comment]
Proposed fix for #1141

**Comment by Hedingber** [comment]
Hi @hayesgb 
Thanks a lot for the contribution!
I asked @theSaarco to review it, in the meanwhile, I see that the lint CI failed, 
Did you follow the instructions on https://github.com/mlrun/mlrun/blob/development/CONTRIBUTING.md and specifically running `make fmt` and `make lint` ?
Let me know if you're having any trouble with this

**Comment by theSaarco** [comment]
@hayesgb, thank you for contributing.
In terms of testing, we have a manual integration test located in `tests/integration/azure_blob/test_azure_blob.py` - these tests are manual since they require you to provide access parameters for the Azure storage account you wish to test against. These details are given in a `yml` file in the same directory. 
Given that you've added new access modes which use different parameters, please add the option to specify these parameters in this test, so it can be used to test your new additions. Please let me know if you need assistance for this.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:81]: Why always use `start=0`? You're ignoring the `offset` and `size` parameters here.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:119]: Why are you filtering for directories? The `listdir` API is expected to return both files and directories under the specified path.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:102]: If I understand correctly, if this is called on a directory path, the results will be unpredictable - `ls` will be successful on a directory path (unlike the call to `get_blob_client` used while using `bsc` which would fail with `does not exist`), and the stats returned will reflect those of an arbitrary first file from the directory listing. This is not the expected behavior - this call should make sure the `key` actually points to a file.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:87]: I just noticed that this comment is not correct, as we're using `overwrite=True`, so it will not fail but rather overwrite. This is of course not your fault 😃. 
Still, I'd prefer in general for this method to blow up if `append=True` is provided since it may cause data loss to the user when expecting an append operation and getting an overwrite instead. 
If you could add this test, that's great. Otherwise I'll do that on a different PR.

**Review by theSaarco — CHANGES_REQUESTED** [review]

File: mlrun/datastore/azure_blob.py
Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:81]: self._filesystem.cat_file() calculates the mlrun's size parameter, instead of taking it explicitly.  That being the case, the concern would be that if start is allowed to be anything other than zero, we could create reading overlapping fragments.

I could update adlfs to accept size in a future release to accept size and set that to length.  Thoughts?

File: mlrun/datastore/azure_blob.py
Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:87]: I just really thought about this operation.  It's worth noting that a BlockBlob (the default in Azure) does not support append operations at all.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [reply mlrun/datastore/azure_blob.py:81]: Yes, I see that `cat_file` indeed accepts `start` and `end` rather than a `size`. It does feel a bit awkward to do a translation just for `adlfs` to reverse the calculation. In this context, adding an option to `adlfs` to accept size will make this better (though not compliant with the `fsspec` default interface).
Still, I'm not sure I follow the overlap concern - why would this translation result in overlaps? Are you concerned about a `size` that exceeds object size?

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [reply mlrun/datastore/azure_blob.py:87]: Yes, that's why it was omitted in the basic implementation - you need to use AppendBlob type (which is not always supported) to work with append operations. Of course, I'm less concerned about not supporting append - I'm concerned about not notifying the user that append in implicitly replaced with overwrite. That's why failing such a call seems appropriate.

File: mlrun/datastore/azure_blob.py
Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:119]: Updated.

File: mlrun/datastore/azure_blob.py
Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:102]: Updated to test that key is a file and introduced error handling

File: mlrun/datastore/azure_blob.py
Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:81]: Size exceeding object_size would be a concern.  Another would be multiple calls, where start is zero, then non-zero.

If I add a size parameter to cat_file, I would handle it in kwargs

Comment (theSaarco) [reply mlrun/datastore/azure_blob.py:81]: I'm probably misunderstanding your concern here - it's legitimate for a client to generate reads over overlapping regions by specifying overlapping `offset`/`size` parameters. Usually that wouldn't be the case, as you'd use a fixed `size` and loop with `offset = prev_offset + size` until you reach EOF, but it's definitely not the intent to block other usages.

Comment (hayesgb) [reply mlrun/datastore/azure_blob.py:81]: I'll pass offset in to start, since it looks like that's a requirement to pass the integration tests.

File: tests/integration/azure_blob/test_azure_blob.py
Comment (theSaarco) [inline tests/integration/azure_blob/test_azure_blob.py:37]: In the `azure_connection_configured` function you're testing for all auth methods parameters to be configured. I think this is a bit of an overkill - if I just want to test a single method, this will not run the test unless I configure all the auth methods - SAS token, secrets, connection string etc. 
I propose to configure it differently, something like this:

```suggestion
AUTH_METHODS_AND_REQUIRED_PARAMS = {
    "conn_str": [
        "AZURE_STORAGE_CONNECTION_STRING"
    ],
    "sas_token": [
        "AZURE_STORAGE_ACCOUNT_NAME", "AZURE_STORAGE_SAS_TOKEN"
    ],
    "account_key": [
        "AZURE_STORAGE_ACCOUNT_NAME", "AZURE_STORAGE_ACCOUNT_KEY"
    ],
    "spn": [
        "AZURE_STORAGE_ACCOUNT_NAME",
        "AZURE_STORAGE_CLIENT_ID",
        "AZURE_STORAGE_CLIENT_SECRET",
        "AZURE_STORAGE_TENANT_ID",
    ]
}
```

Then in the `azure_connection_configured` method you can just verify that a single auth method got all of the parameters needed for it. 
In the `prepare_env` implementation you just need to iterate over the parameters in the method selected and assign the OS environment from the config (assuming the params are configured, of course). If params are not configured for a given auth method, the `prepare_env` can return an indication to that, and the test will skip running that method.

File: tests/integration/azure_blob/test_azure_blob.py
Comment (theSaarco) [inline tests/integration/azure_blob/test_azure_blob.py:80]: See my comment above.

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:111]:
```suggestion
            if len(files) == 1 and files[0]["kind"] == "file":
```

File: mlrun/datastore/azure_blob.py
Comment (theSaarco) [inline mlrun/datastore/azure_blob.py:115]:
```suggestion
            elif len(files) == 1 and files[0]["kind"] == "directory":
                raise FileNotFoundError("Operation expects a file not a directory!")
```

Are we guaranteed that when doing an `ls` operation on a directory, the first item returned will be the directory itself? I would actually expect an `ls` operation to not return the directory at all (in Linux it may return `.` in some cases, but I don't know how this works for `adlfs`).

I think it might be enough in this case to simply fail in this case:
```suggestion
            else:
                raise FileNotFoundError("Operation expects a single file, not a directory!")
```

Since the only valid case is covered before.

**Review by theSaarco — CHANGES_REQUESTED** [review]

File: tests/integration/azure_blob/test_azure_blob.py
Comment (theSaarco) [reply tests/integration/azure_blob/test_azure_blob.py:37]: Coming to think of it - if doing this then can completely remove the `@pytest.mark.skipif` in the beginning of the tests - simply iterate over all auth methods, and only perform the test if params are configured for it. If none of the methods are configured, the test will succeed by default.

File: tests/integration/azure_blob/test_azure_blob.py
Comment (hayesgb) [reply tests/integration/azure_blob/test_azure_blob.py:43]: Should this be configured to assert that, at the very least, the "AZURE_STORAGE_CONNECTION_STRING" is configured?

Comment (theSaarco) [reply tests/integration/azure_blob/test_azure_blob.py:43]: No. The point is that this test is opt-in, since it requires manual configuration and is designed to run manually. However, all tests under `tests/integration` are included in the integration test suite. When running the overall suite of tests (for example in the CI when doing a PR) we don't have manual configurations, so the test needs to either pass silently or be skipped. If failing the test, it will fail the test suite and for example block your PR.

Comment (hayesgb) [reply tests/integration/azure_blob/test_azure_blob.py]: I reverted this back to the original function, as implemented, which expects AZURE_STORAGE_CONNECTION_STRING.

Are you saying you'd prefer to eliminate this function, and the `pytest.mark.skipif` calls?

**Comment by hayesgb** [comment]
As a side-note, when I try to run `make test-dockerized` on my local machine, I'm getting the following error:

`#5 1.083 E: The repository 'http://security.debian.org/debian-security buster/updates InRelease' is not signed.
#5 1.083 W: GPG error: http://deb.debian.org/debian buster-updates InRelease: At least one invalid signature was encountered.
#5 1.083 E: The repository 'http://deb.debian.org/debian buster-updates InRelease' is not signed.
`

Do you recommend I assume this is trusted?

**Comment by Hedingber** [comment]
@hayesgb I've never encountered this error, googling it brought up this one https://stackoverflow.com/questions/62473932/atleast-one-invalid-signature-was-encountered
I would start with the suggestions there (Cleaning caches and images) before manually "making it trusted"
BTW, note that the reason CI doesn't run automatically on all of your commits (it's running only when I approve) is because you're a first time contributor, once we'll merge one of the PRs, it will be more comfortable.
Another thing on that regard, CI passed successfully 🥳

**Comment by hayesgb** [comment]
Thanks @Hedingber -- Do I need to make any additional changes at this point?  Can we consider the `azure_connection_configured()` request closed?

**Comment by theSaarco** [comment]
@hayesgb - I'm fine with the changes done so far. I have some ideas for improving the test further, but no reason to block this PR for them at this point. 🎉

**Review by theSaarco — APPROVED** [review]

**Comment by Hedingber** [comment]
@hayesgb 
PR was merged, you're officially an MLRun contributor now 🎉

**Pull request merged**
